src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -1504,6 +1504,422 @@ public static void Main()
+                // (11,21): warning CS8602: Dereference of a possibly null reference.
+                //         if (flag) { b.X.ToString(); } // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.X").WithLocation(11, 21),
+                // (16,21): warning CS8602: Dereference of a possibly null reference.
+                //         if (flag) { b.X.ToString(); } // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.X").WithLocation(16, 21));
+        }
+
+        [Fact(Skip = "PROTOTYPE(records): positional property attributes")]
+        public void WithExpr_NullableAnalysis_07()
+        {
+            var src = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+data class B([AllowNull] string X)
+{
+    public B Clone() => new B(X);
+
+    static void M1(B b)
+    {
+        b.X.ToString();
+        b = b with { X = null };
+        b.X.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(new[] { src, AllowNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                //         b.X.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.X").WithLocation(13, 9));
+        }
+
+        [Fact]
+        public void WithExpr_NullableAnalysis_08()
+        {
+            var src = @"
+#nullable enable
+data class B(string? X, string? Y)
+{
+    public B Clone() => new B(X, Y);
+
+    static void M1(B b1)
+    {
+        B b2 = b1 with { X = ""hello"" };
+        B b3 = b1 with { Y = ""world"" };
+        B b4 = b2 with { Y = ""world"" };
+        
+        b1.X.ToString(); // 1
+        b1.Y.ToString(); // 2
+        b2.X.ToString();
+        b2.Y.ToString(); // 3
+        b3.X.ToString(); // 4
+        b3.Y.ToString();
+        b4.X.ToString();
+        b4.Y.ToString();
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                //         b1.X.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b1.X").WithLocation(13, 9),
+                // (14,9): warning CS8602: Dereference of a possibly null reference.
+                //         b1.Y.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b1.Y").WithLocation(14, 9),
+                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                //         b2.Y.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b2.Y").WithLocation(16, 9),
+                // (17,9): warning CS8602: Dereference of a possibly null reference.
+                //         b3.X.ToString(); // 4
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b3.X").WithLocation(17, 9));
+        }
```

Comment:
Could you add some test for region analysis APIs (what data flows in or out of expression)?